### PR TITLE
Assignment2 - Joshua Lester & Spencer West

### PR DESCRIPTION
**Pull request by Joshua-Lester3**
@Joshua-Lester3 - Joshua Lester
@saawest - Spencer West
URL: https://purple-plant-006657c1e.5.azurestaticapps.net/

### CODE REVIEW

**Comment by SergeiAshikhmin** [comment]
I will review this :)

File: wordle-web/layouts/default.vue [hunk 1]
```diff
@@ -0,0 +1,52 @@
+<template>
+  <v-app>
+    <v-app-bar color="primary">
+      <template v-slot:prepend>
+        <div class="text-h5 ml-5">
+          <v-icon
+            color="secondary"
+            @click="router.push('/index')"
```
Comment (SergeiAshikhmin) [inline wordle-web/layouts/default.vue]: I think it is a bug because when I click the icon, it sends me to the unfound page. I think you just forgot to remove /index
- Clicking the name and icon of the app take you to the home page ❌

**Review by SergeiAshikhmin — COMMENTED** [review]
Good job 👍 
I like how your project looks like and how you structure your code. There are several bugs like the drawer opening when you resize the page or the web app doesn't load in the correct size when refreshed. But I don't know if we have to fix this.

Also, nice theme implementation. You did it way cleaner than I did. 

1. Create a custom component to show available words ✅
- Create/extend a word list component with a method called validWords that returns an array of valid words based on current guesses ✅
- Create unit tests for the validWords method ✅
- Display the number of valid words based on the entered letters ✅
- When the number is clicked, show a dialog with all the available words with scrolling if there are too many to show on the screen. ✅
- When a word is clicked, the word should be entered as the guess but not submitted ✅
- Add a hover over for the word so it is obvious that it can be clicked. (No CSS) ✅

2. Convert the sidebar to an App Bar ✅
- Name of the app on the left with an icon ✅
- Choose an icon for the app (from Material Design or somewhere on the web) ✅
- Clicking the name and icon of the app take you to the home page ❌
- Hamburger menu on the right ✅
- Menu option for an About page ✅
  - Create an About page that has a few sentences about this class project and how great Meg is ✅
  - Make sure this has a back button ✅

3. Add a settings dialog ✅
- Add a settings selection to the App Bar's menu with a gear icon. This should open this dialog regardless of where you are on the site ✅
- Option for dark and light mode ✅
- Develop two additional color schemes (with creative names) and allow the user to change to them. Schemes should look good in both light and dark mode ✅
- The above two items must be implemented with built in Veutify features ✅

4. Style the main game page ✅
- Make it look like [wordle.com](https://www.nytimes.com/games/wordle/index.html) or the Wordle mobile app ✅
- Add gradient to guesses blocks and keys (with CSS) ✅
- Add drop shadow to keys (No CSS) ✅
- Make the app responsive (No CSS) ✅

5. Include a link to your Azure App in your PR comments.

## Extra Credit

Didn't find any extra credit

File: wordle-web/layouts/default.vue (hunk 1 above)
Comment (Joshua-Lester3) [reply wordle-web/layouts/default.vue]: Thanks for the catch Sergei! I was debugging something else and added that and forgot to remove it. Fixed it.

**Comment by connerhanson** [comment]
I will review this as-well.

File: wordle-web/components/HelpDialog.vue [hunk 2]
```diff
@@ -19,16 +19,7 @@
...
-          v-for="letter in exampleWord.letters"
-          :key="letter.char"
-          width="50px"
-          height="50px"
-          :color="letter.color"
-          class="mx-1 text-center"
-        >
-          <v-card-text> {{ letter.char }} </v-card-text>
-        </v-card>
+        <GameBoardGuess :guess="exampleWord" />
```
Comment (connerhanson) [inline wordle-web/components/HelpDialog.vue:22]: I like the way you simplified this.

File: wordle-web/components/SettingsDialog.vue [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+const modelValue = defineModel<boolean>({ default: false });
+const theme = useTheme();
+const selectedTheme = ref('Regular');
+
+onMounted(() => {
+  var defaultTheme = nuxtStorage.localStorage.getData('theme');
+  theme.global.name.value = defaultTheme ?? 'regularDark';
+});
+
+function toggleTheme() {
```
Comment (connerhanson) [inline wordle-web/components/SettingsDialog.vue:51]: great job on getting the themes to work on either light or dark, this was one of the hardest parts for me.

File: wordle-web/components/ValidWords.vue [hunk 4]
```diff
@@ -0,0 +1,98 @@
+<template>
+  <v-dialog v-model="show">
+    <v-card class="ma-auto" min-width="300">
+      <v-sheet :color="validWords.length === 0 ? 'unknown' : 'secondary'">
+        <v-card-title class="text-center"
+          >Valid Words
+          <v-btn
```
Comment (connerhanson) [inline wordle-web/components/ValidWords.vue:7]: Good implementation on the button with an infinite scroll sheet, I used a select menu, but your implementation looks much prettier.

File: wordle-web/scripts/theme.ts [hunk 5]
```diff
@@ -0,0 +1,12 @@
+export class ThemeUtils {
+  static help() {}
+}
+
+export enum Theme {
```
Comment (connerhanson) [inline wordle-web/scripts/theme.ts:5]: I like this implementation of enums for the themes. Makes the code a lot more readable for the settings page.

File: wordle-web/scripts/validWordsUtils.ts [hunk 6]
```diff
@@ -0,0 +1,168 @@
...
+        lettersIndex++
+      ) {
+        this.letters[lettersIndex].numberDiscoveredInWord =
+          this.letters[lettersIndex].runningTotalDiscovered;
+        this.letters[lettersIndex].runningTotalDiscovered = 0;
+      }
+    }
+  }
+}
+
```
Comment (connerhanson) [inline wordle-web/scripts/validWordsUtils.ts:144]: I like the implementation for this, it took me a while to understand it but seems super robust.

File: wordle-web/tests/wordList.test.ts [hunk 7]
```diff
@@ -0,0 +1,142 @@
...
+  'WHICH',
+];
+
+const wordList2 = ['ABACA', 'ABODE', 'TAROT'];
+
+describe('validWords_enteredLettersIn_returnsCorrectWordAndLength', () => {
+  it('enteredA_returnsFirstTwoWords', () => {
+    let game = new Game();
+    game.secretWord = 'HELLO';
+    game.addLetter('A');
```
Comment (connerhanson) [inline wordle-web/tests/wordList.test.ts:22]: Great test, the only thing I might recommend is using the whole word List rather than a short one, it may help show any bugs in the valid Word method if there is any.

**Review by connerhanson — COMMENTED** [review]
Overall great job it was hard to find anything that really needed improvement. 

1. Create a custom component to show available words ✅
- Create/extend a word list component with a method called validWords that returns an array of valid words based on current guesses ✅
- Create unit tests for the validWords method ✅
- Display the number of valid words based on the entered letters ✅
- When the number is clicked, show a dialog with all the available words with scrolling if there are too many to show on the screen. ✅  
- When a word is clicked, the word should be entered as the guess but not submitted ✅
- Add a hover over for the word so it is obvious that it can be clicked. (No CSS) ✅

2. Convert the sidebar to an App Bar ✅
- Name of the app on the left with an icon ✅
- Choose an icon for the app (from Material Design or somewhere on the web) ✅
- Clicking the name and icon of the app take you to the home page ✅
- Hamburger menu on the right ✅
- Menu option for an About page ✅
  - Create an About page that has a few sentences about this class project and how great Meg is ✅
  - Make sure this has a back button ✅

3. Add a settings dialog ✅
- Add a settings selection to the App Bar's menu with a gear icon. This should open this dialog regardless of where you are on the site ✅ 
- Option for dark and light mode ✅
- Develop two additional color schemes (with creative names) and allow the user to change to them. Schemes should look good in both light and dark mode ✅
- The above two items must be implemented with built in Veutify features ✅

4. Style the main game page✅
- Make it look like [wordle.com](https://www.nytimes.com/games/wordle/index.html) or the Wordle mobile app ✅
- Add gradient to guesses blocks and keys (with CSS) ✅
- Add drop shadow to keys (No CSS) ✅
- Make the app responsive (No CSS) ✅

File: wordle-web/components/HelpDialog.vue (hunk 2 above)
Comment (Joshua-Lester3) [reply wordle-web/components/HelpDialog.vue:22]: Thank you!

File: wordle-web/components/SettingsDialog.vue (hunk 3 above)
Comment (Joshua-Lester3) [reply wordle-web/components/SettingsDialog.vue:51]: Definitely was tough. Thanks :)

File: wordle-web/components/ValidWords.vue (hunk 4 above)
Comment (Joshua-Lester3) [reply wordle-web/components/ValidWords.vue:7]: Thank you! I would love to implement a search feature like I saw others do.

File: wordle-web/scripts/theme.ts (hunk 5 above)
Comment (Joshua-Lester3) [reply wordle-web/scripts/theme.ts:5]: Thanks :)

File: wordle-web/scripts/validWordsUtils.ts (hunk 6 above)
Comment (Joshua-Lester3) [reply wordle-web/scripts/validWordsUtils.ts:144]: Yeah, I'm not sure if it's the prettiest code, but it works.

File: wordle-web/tests/wordList.test.ts (hunk 7 above)
Comment (Joshua-Lester3) [reply wordle-web/tests/wordList.test.ts:22]: True. I chose to do this because I can simulate anything with a smaller list. Sifting through the larger list to find what to look for is too time consuming for me. This way, I could spend more time thinking of edge cases! :)

File: wordle-web/tests/wordList.test.ts [hunk 8]
```diff
@@ -0,0 +1,142 @@
...
+    let validWords = utils.validWords();
+    expect(validWords.length).toBe(1);
+    expect(validWords[0]).toBe('AARGH');
+  });
+  it('noLettersEntered_returnsFullList', () => {
+    let game = new Game();
+    game.secretWord = 'HELLO';
+    let utils = new ValidWordsUtils(game, wordList);
+    let validWords = utils.validWords();
+    expect(validWords.length).toBe(7);
```
Comment (vicdotexe) [inline wordle-web/tests/wordList.test.ts:44]:
```suggestion
    expect(validWords.length).toBe(wordList.length);
```
Hardcoding 7 works fine because we know the length of the list at time of writing, however we can guard against future changes here in addition to making the assert portion read a bit clearer: "expect validWords to have the same count as the wordList"

File: wordle-web/scripts/validWordsUtils.ts [hunk 9]
```diff
@@ -0,0 +1,168 @@
...
+      let guess = this.game.guesses[guessesIndex];
+      this.setRunningTotalForAllLettersToZero();
+      if (!guess.isFilled) {
+        return;
+      }
+      let lettersArray = guess.letters;
+      // Adjust letter helpers based on correct, misplaced, or wrong states
+      for (let index = 0; index < lettersArray.length; index++) {
+        let lettersIndex = lettersArray[index].char.charCodeAt(0) - 65;
+        if (lettersArray[index].state === LetterState.Correct) {
```
Comment (vicdotexe) [inline wordle-web/scripts/validWordsUtils.ts:108]:
```suggestion
        switch (lettersArray[index].state){
          case LetterState.Correct:
```
Breaking this `else if` chain up into a switch statement would help keep things a bit more organized and clear.

**Review by vicdotexe — COMMENTED** [review]
1. Create a custom component to show available words 
- Create/extend a word list component with a method called validWords that returns an array of valid words based on current guesses ✅
- Create unit tests for the validWords method ✅
- Display the number of valid words based on the entered letters ✅
- When the number is clicked, show a dialog with all the available words with scrolling if there are too many to show on the screen. ✅
- When a word is clicked, the word should be entered as the guess but not submitted ✅
- Add a hover over for the word so it is obvious that it can be clicked. (No CSS) ✅

2. Convert the sidebar to an App Bar 
- Name of the app on the left with an icon ✅
- Choose an icon for the app (from Material Design or somewhere on the web) ✅
- Clicking the name and icon of the app take you to the home page ✅
- Hamburger menu on the right ✅
- Menu option for an About page ✅
  - Create an About page that has a few sentences about this class project and how great Meg is ✅
  - Make sure this has a back button ✅

3. Add a settings dialog 
- Add a settings selection to the App Bar's menu with a gear icon. This should open this dialog regardless of where you are on the site ✅
- Option for dark and light mode ✅
- Develop two additional color schemes (with creative names) and allow the user to change to them. Schemes should look good in both light and dark mode ✅
- The above two items must be implemented with built in Veutify features ✅

4. Style the main game page 
- Make it look like [wordle.com](https://www.nytimes.com/games/wordle/index.html) or the Wordle mobile app ✅
- Add gradient to guesses blocks and keys (with CSS) ✅
- Add drop shadow to keys (No CSS) ✅
- Make the app responsive (No CSS) ✅

5. Include a link to your Azure App in your PR comments. ✅

## Extra Credit

- Add an appropriate sound when the UI letter buttons are clicked (5 points) ✅
- Add creative styling following good UI/UX practices 
  - Look at products by Google or some of the vuetify components for good UX/UI ideas
  - https://m2.material.io/design/guidelines-overview
  
> [!NOTE]
> Overall, great job, this has met all the acceptance criteria for the assignment. You went in depth with your implementation to calculate the valid words, which worked great, however I believe we only need to be calculating possible words based on entered guesses, rather than every keystroke on partial guesses. My only comments would be:
> - UX: it would be nice if the dialog closed when a word was picked from the list, and it would be nice if the theme changed without needing to press submit.
> - Code: it was clean for the most part, perhaps try to use `const` over `let` when we know a variable won't be changing.
